src/libcrun/linux.c

```diff
@@ -1043,6 +1043,75 @@ enum
   LABEL_XATTR,
 };
 
+static int do_mount (libcrun_container_t *container, const char *source, int targetfd,
+                     const char *target, const char *fstype, unsigned long mountflags,
+                     const void *data, int label_how, libcrun_error_t *err);
+
+static bool
+has_mount_for (libcrun_container_t *container, const char *destination)
+{
+  size_t i;
+  runtime_spec_schema_config_schema *def = container->container_def;
+
+  for (i = 0; i < def->mounts_len; i++)
+    {
+      if (strcmp (def->mounts[i]->destination, destination) == 0)
+        return true;
+    }
+  return false;
+}
+
+static int
+do_masked_or_readonly_path (libcrun_container_t *container, const char *rel_path, bool readonly,
+                            libcrun_error_t *err)
+{
+  size_t rootfs_len = get_private_data (container)->rootfs_len;
+  const char *rootfs = get_private_data (container)->rootfs;
+  int rootfsfd = get_private_data (container)->rootfsfd;
+  cleanup_close int pathfd = -1;
+  int ret;
+  mode_t mode;
+
+  if (rel_path[0] == '/')
+    rel_path++;
+
+  pathfd = safe_openat (rootfsfd, rootfs, rootfs_len, rel_path, O_PATH | O_CLOEXEC, 0, err);
+  if (UNLIKELY (pathfd < 0))
+    {
+      if (errno != ENOENT && errno != EACCES)
+        return crun_make_error (err, errno, "open `%s`", rel_path);
+
+      crun_error_release (err);
+      return 0;
+    }
+
+  if (readonly)
+    {
+      char source_buffer[64];
+      sprintf (source_buffer, "/proc/self/fd/%d", pathfd);
+
+      ret = do_mount (container, source_buffer, pathfd, rel_path, NULL, MS_BIND | MS_PRIVATE | MS_RDONLY | MS_REC, NULL,
+                      LABEL_NONE, err);
+      if (UNLIKELY (ret < 0))
+        return ret;
+    }
+  else
+    {
+      ret = get_file_type_fd (pathfd, &mode);
+      if (UNLIKELY (ret < 0))
+        return ret;
+
+      if ((mode & S_IFMT) == S_IFDIR)
+        ret = do_mount (container, "tmpfs", pathfd, rel_path, "tmpfs", MS_RDONLY, "size=0k", LABEL_MOUNT, err);
+      else
+        ret = do_mount (container, "/dev/null", pathfd, rel_path, NULL, MS_BIND | MS_UNBINDABLE | MS_REC, NULL,
```

Comment:
Needs MS_RDONLY ?

Reply:
And MS_REC can be dropped

Reply:
not sure MS_RDONLY makes any difference with a masked path but I'll add it.  Thinking more of it, it probably doesn't need MS_UNBINDABLE as well, I'll simplify it to "MS_BIND | MS_RDONLY".

Thanks for the suggestion!